File: backend/src/morado/schemas/test_case.py
"""Test Case Pydantic schemas.

This module provides schemas for test case API request/response validation.
"""

from enum import Enum

from pydantic import BaseModel, ConfigDict, Field

from morado.schemas.common import PaginatedResponse, TimestampMixin, UUIDMixin


class TestCasePriority(str, Enum):
    """测试用例优先级"""

    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    CRITICAL = "critical"


class TestCaseStatus(str, Enum):
    """测试用例状态"""

    DRAFT = "draft"
    ACTIVE = "active"
    DEPRECATED = "deprecated"
    ARCHIVED = "archived"


class TestCaseBase(BaseModel):
    """测试用例基础schema

    包含测试用例的基本字段，用于创建和更新操作。
    """

    name: str = Field(min_length=1, max_length=200, description="用例名称")
    description: str | None = Field(default=None, description="用例描述")
    priority: TestCasePriority = Field(
        default=TestCasePriority.MEDIUM, description="优先级"
    )
    status: TestCaseStatus = Field(default=TestCaseStatus.DRAFT, description="状态")
    category: str | None = Field(default=None, max_length=100, description="分类")
    tags: list[str] | None = Field(default=None, description="标签")
    preconditions: str | None = Field(default=None, description="前置条件")
    postconditions: str | None = Field(default=None, description="后置条件")
    execution_order: str = Field(
        default="sequential", pattern="^(sequential|parallel)$", description="执行顺序"
    )
    timeout: int = Field(default=300, ge=1, description="超时时间（秒）")
    retry_count: int = Field(default=0, ge=0, description="重试次数")
    continue_on_failure: bool = Field(default=False, description="失败时是否继续")
    test_data: dict | None = Field(default=None, description="测试数据")
    environment: str = Field(default="test", max_length=20, description="执行环境")
    version: str = Field(default="1.0.0", max_length=20, description="版本号")
    is_automated: bool = Field(default=True, description="是否自动化")

    model_config = ConfigDict(
        from_attributes=True,
        json_schema_extra={
            "example": {
                "name": "用户登录测试",
                "description": "测试用户登录功能",
                "priority": "high",
                "status": "active",
                "category": "用户管理",
                "tags": ["登录", "认证"],
                "timeout": 300,
                "environment": "test",
            }
        },
    )


class TestCaseCreate(TestCaseBase):
    """创建测试用例schema

    用于创建新的测试用例。
    """

    created_by: int | None = Field(default=None, description="创建者ID")


class TestCaseUpdate(BaseModel):
    """更新测试用例schema

    所有字段都是可选的，只更新提供的字段。
    """

    name: str | None = Field(
        default=None, min_length=1, max_length=200, description="用例名称"
    )
    description: str | None = Field(default=None, description="用例描述")
    priority: TestCasePriority | None = Field(default=None, description="优先级")
    status: TestCaseStatus | None = Field(default=None, description="状态")
    category: str | None = Field(default=None, max_length=100, description="分类")
    tags: list[str] | None = Field(default=None, description="标签")
    preconditions: str | None = Field(default=None, description="前置条件")
    postconditions: str | None = Field(default=None, description="后置条件")
    execution_order: str | None = Field(
        default=None, pattern="^(sequential|parallel)$", description="执行顺序"
    )
    timeout: int | None = Field(default=None, ge=1, description="超时时间（秒）")
    retry_count: int | None = Field(default=None, ge=0, description="重试次数")
    continue_on_failure: bool | None = Field(default=None, description="失败时是否继续")
    test_data: dict | None = Field(default=None, description="测试数据")
    environment: str | None = Field(default=None, max_length=20, description="执行环境")
    version: str | None = Field(default=None, max_length=20, description="版本号")
    is_automated: bool | None = Field(default=None, description="是否自动化")

    model_config = ConfigDict(from_attributes=True)


class TestCaseResponse(TestCaseBase, TimestampMixin, UUIDMixin):
    """测试用例响应schema

    包含完整的测试用例信息，包括ID和时间戳。
    """

    id: int = Field(description="用例ID")
    created_by: int | None = Field(default=None, description="创建者ID")

    model_config = ConfigDict(
        from_attributes=True,
        json_schema_extra={
            "example": {
                "id": 1,
                "uuid": "550e8400-e29b-41d4-a716-446655440000",
                "name": "用户登录测试",
                "description": "测试用户登录功能",
                "priority": "high",
                "status": "active",
                "category": "用户管理",
                "tags": ["登录", "认证"],
                "timeout": 300,
                "environment": "test",
                "version": "1.0.0",
                "is_automated": True,
                "created_by": 1,
                "created_at": "2024-01-01T00:00:00Z",
                "updated_at": "2024-01-01T00:00:00Z",
            }
        },
    )


class TestCaseListResponse(PaginatedResponse[TestCaseResponse]):
    """测试用例列表响应schema"""


# Test Case Script Schemas


class TestCaseScriptBase(BaseModel):
    """测试用例-脚本关联基础schema"""

    test_case_id: int = Field(description="测试用例ID")
    script_id: int = Field(description="脚本ID")
    execution_order: int = Field(default=0, description="执行顺序")
    is_enabled: bool = Field(default=True, description="是否启用")
    script_parameters: dict | None = Field(default=None, description="脚本参数覆盖")
    description: str | None = Field(default=None, description="说明")

    model_config = ConfigDict(from_attributes=True)


class TestCaseScriptCreate(TestCaseScriptBase):
    """创建测试用例-脚本关联schema"""


class TestCaseScriptResponse(TestCaseScriptBase, TimestampMixin):
    """测试用例-脚本关联响应schema"""

    id: int = Field(description="关联ID")

    model_config = ConfigDict(from_attributes=True)


# Test Case Component Schemas


class TestCaseComponentBase(BaseModel):
    """测试用例-组件关联基础schema"""

    test_case_id: int = Field(description="测试用例ID")
    component_id: int = Field(description="组件ID")
    execution_order: int = Field(default=0, description="执行顺序")
    is_enabled: bool = Field(default=True, description="是否启用")
    component_parameters: dict | None = Field(default=None, description="组件参数覆盖")
    description: str | None = Field(default=None, description="说明")

    model_config = ConfigDict(from_attributes=True)


class TestCaseComponentCreate(TestCaseComponentBase):
    """创建测试用例-组件关联schema"""


class TestCaseComponentResponse(TestCaseComponentBase, TimestampMixin):
    """测试用例-组件关联响应schema"""

    id: int = Field(description="关联ID")

    model_config = ConfigDict(from_attributes=True)

File: frontend/src/api/test-case.ts
/**
 * Test Case API Client
 *
 * Layer 4: Test Cases - Test Case Management and Execution
 * Provides functions for managing test cases that reference scripts and components.
 */

import apiClient, { type PaginatedResponse } from './index';

// Test Case Types
export type TestCasePriority = 'low' | 'medium' | 'high' | 'critical';
export type TestCaseStatus = 'draft' | 'active' | 'deprecated' | 'archived';

export interface TestCase {
  id: number;
  uuid: string;
  name: string;
  description?: string;
  priority: TestCasePriority;
  status: TestCaseStatus;
  category?: string;
  tags?: string[];
  preconditions?: string;
  postconditions?: string;
  executionOrder: string;
  timeout: number;
  timeoutOverride?: number;
  retryCount: number;
  continueOnFailure: boolean;
  testData?: Record<string, unknown>;
  environment: string;
  version: string;
  isAutomated: boolean;
  scriptCount?: number;
  componentCount?: number;
  lastExecutionAt?: string;
  lastExecutionStatus?: 'passed' | 'failed' | 'skipped';
  createdBy?: number;
  createdAt: string;
  updatedAt: string;
}

export interface TestCaseCreate {
  name: string;
  description?: string;
  priority?: TestCasePriority;
  status?: TestCaseStatus;
  category?: string;
  tags?: string[];
  preconditions?: string;
  postconditions?: string;
  executionOrder?: string;
  timeout?: number;
  retryCount?: number;
  continueOnFailure?: boolean;
  testData?: Record<string, unknown>;
  environment?: string;
  version?: string;
  isAutomated?: boolean;
  createdBy?: number;
}

export interface TestCaseUpdate {
  name?: string;
  description?: string;
  priority?: TestCasePriority;
  status?: TestCaseStatus;
  category?: string;
  tags?: string[];
  preconditions?: string;
  postconditions?: string;
  executionOrder?: string;
  timeout?: number;
  retryCount?: number;
  continueOnFailure?: boolean;
  testData?: Record<string, unknown>;
  environment?: string;
  version?: string;
  isAutomated?: boolean;
}

export interface TestCaseScript {
  id: number;
  testCaseId: number;
  scriptId: number;
  executionOrder: number;
  isEnabled: boolean;
  scriptParameters?: Record<string, unknown>;
  description?: string;
  createdAt: string;
  updatedAt: string;
}

export interface TestCaseScriptCreate {
  testCaseId: number;
  scriptId: number;
  executionOrder?: number;
  isEnabled?: boolean;
  scriptParameters?: Record<string, unknown>;
  description?: string;
}

export interface TestCaseScriptUpdate {
  executionOrder?: number;
  isEnabled?: boolean;
  scriptParameters?: Record<string, unknown>;
  description?: string;
}

export interface TestCaseComponent {
  id: number;
  testCaseId: number;
  componentId: number;
  executionOrder: number;
  isEnabled: boolean;
  componentParameters?: Record<string, unknown>;
  description?: string;
  createdAt: string;
  updatedAt: string;
}

export interface TestCaseComponentCreate {
  testCaseId: number;
  componentId: number;
  executionOrder?: number;
  isEnabled?: boolean;
  componentParameters?: Record<string, unknown>;
  description?: string;
}

export interface TestCaseComponentUpdate {
  executionOrder?: number;
  isEnabled?: boolean;
  componentParameters?: Record<string, unknown>;
  description?: string;
}

export interface TestCaseListParams {
  page?: number;
  pageSize?: number;
  priority?: TestCasePriority;
  status?: TestCaseStatus;
  category?: string;
  isAutomated?: boolean;
  search?: string;
  tags?: string[];
  environment?: string;
}

export interface TestCaseExecutionResult {
  success: boolean;
  testCaseId: number;
  executionId: number;
  scriptResults: Array<{
    scriptId: number;
    scriptName: string;
    success: boolean;
    duration: number;
    error?: string;
  }>;
  componentResults: Array<{
    componentId: number;
    componentName: string;
    success: boolean;
    duration: number;
    error?: string;
  }>;
  totalDuration: number;
  timestamp: string;
  environment: string;
}

// Test Case API Functions

/**
 * Get list of test cases with pagination and filtering
 */
export async function getTestCases(
  params?: TestCaseListParams,
): Promise<PaginatedResponse<TestCase>> {
  const response = await apiClient.get<PaginatedResponse<TestCase>>('/test-cases', { params });
  return response.data;
}

/**
 * Get a single test case by ID
 */
export async function getTestCase(id: number): Promise<TestCase> {
  const response = await apiClient.get<TestCase>(`/test-cases/${id}`);
  return response.data;
}

/**
 * Create a new test case
 */
export async function createTestCase(data: TestCaseCreate): Promise<TestCase> {
  const response = await apiClient.post<TestCase>('/test-cases', data);
  return response.data;
}

/**
 * Update an existing test case
 */
export async function updateTestCase(id: number, data: TestCaseUpdate): Promise<TestCase> {
  const response = await apiClient.patch<TestCase>(`/test-cases/${id}`, data);
  return response.data;
}

/**
 * Delete a test case
 */
export async function deleteTestCase(id: number): Promise<void> {
  await apiClient.delete(`/test-cases/${id}`);
}

/**
 * Duplicate a test case
 */
export async function duplicateTestCase(id: number, newName?: string): Promise<TestCase> {
  const response = await apiClient.post<TestCase>(`/test-cases/${id}/duplicate`, { name: newName });
  return response.data;
}

/**
 * Execute a test case
 */
export async function executeTestCase(
  id: number,
  runtimeParams?: Record<string, unknown>,
): Promise<TestCaseExecutionResult> {
  const response = await apiClient.post<TestCaseExecutionResult>(`/test-cases/${id}/execute`, {
    runtimeParameters: runtimeParams,
  });
  return response.data;
}

/**
 * Get test cases by priority
 */
export async function getTestCasesByPriority(priority: TestCasePriority): Promise<TestCase[]> {
  const response = await apiClient.get<PaginatedResponse<TestCase>>('/test-cases', {
    params: { priority, pageSize: 100 },
  });
  return response.data.items;
}

/**
 * Get test cases by status
 */
export async function getTestCasesByStatus(status: TestCaseStatus): Promise<TestCase[]> {
  const response = await apiClient.get<PaginatedResponse<TestCase>>('/test-cases', {
    params: { status, pageSize: 100 },
  });
  return response.data.items;
}

/**
 * Get test cases by category
 */
export async function getTestCasesByCategory(category: string): Promise<TestCase[]> {
  const response = await apiClient.get<PaginatedResponse<TestCase>>('/test-cases', {
    params: { category, pageSize: 100 },
  });
  return response.data.items;
}

/**
 * Search test cases by name or tags
 */
export async function searchTestCases(query: string): Promise<TestCase[]> {
  const response = await apiClient.get<PaginatedResponse<TestCase>>('/test-cases', {
    params: { search: query, pageSize: 50 },
  });
  return response.data.items;
}

// Test Case Script Association Functions

/**
 * Get scripts associated with a test case
 */
export async function getTestCaseScripts(testCaseId: number): Promise<TestCaseScript[]> {
  const response = await apiClient.get<PaginatedResponse<TestCaseScript>>(
    `/test-cases/${testCaseId}/scripts`,
  );
  return response.data.items;
}

/**
 * Add a script to a test case
 */
export async function addScriptToTestCase(data: TestCaseScriptCreate): Promise<TestCaseScript> {
  const response = await apiClient.post<TestCaseScript>('/test-case-scripts', data);
  return response.data;
}

/**
 * Update a test case-script association
 */
export async function updateTestCaseScript(
  id: number,
  data: TestCaseScriptUpdate,
): Promise<TestCaseScript> {
  const response = await apiClient.patch<TestCaseScript>(`/test-case-scripts/${id}`, data);
  return response.data;
}

/**
 * Remove a script from a test case
 */
export async function removeScriptFromTestCase(id: number): Promise<void> {
  await apiClient.delete(`/test-case-scripts/${id}`);
}

// Test Case Component Association Functions

/**
 * Get components associated with a test case
 */
export async function getTestCaseComponents(testCaseId: number): Promise<TestCaseComponent[]> {
  const response = await apiClient.get<PaginatedResponse<TestCaseComponent>>(
    `/test-cases/${testCaseId}/components`,
  );
  return response.data.items;
}

/**
 * Add a component to a test case
 */
export async function addComponentToTestCase(
  data: TestCaseComponentCreate,
): Promise<TestCaseComponent> {
  const response = await apiClient.post<TestCaseComponent>('/test-case-components', data);
  return response.data;
}

/**
 * Update a test case-component association
 */
export async function updateTestCaseComponent(
  id: number,
  data: TestCaseComponentUpdate,
): Promise<TestCaseComponent> {
  const response = await apiClient.patch<TestCaseComponent>(`/test-case-components/${id}`, data);
  return response.data;
}

/**
 * Remove a component from a test case
 */
export async function removeComponentFromTestCase(id: number): Promise<void> {
  await apiClient.delete(`/test-case-components/${id}`);
}

/**
 * Get test case execution history
 */
export async function getTestCaseExecutionHistory(
  testCaseId: number,
  limit = 10,
): Promise<TestCaseExecutionResult[]> {
  const response = await apiClient.get<{ items: TestCaseExecutionResult[] }>(
    `/test-cases/${testCaseId}/executions`,
    { params: { limit } },
  );
  return response.data.items;
}
